File: llm.py
"""llm.py — one small, dependency-free way to call four kinds of LLM endpoint.

Providers
  anthropic           Claude, Messages API                         (default)
  openai              OpenAI, Chat Completions API
  gemini              Google Gemini, generateContent API
  openai-compatible   any server that speaks the Chat Completions shape: Ollama, LM Studio, vLLM, most gateways

Configuration (environment variables)
  LLM_PROVIDER   anthropic | openai | gemini | openai-compatible          default: anthropic
  LLM_MODEL      model id. Required for every provider except anthropic   (no guessed defaults for other vendors)
  LLM_BASE_URL   endpoint base. Required for openai-compatible (e.g. http://localhost:11434/v1); optional elsewhere
  API key        ANTHROPIC_API_KEY | OPENAI_API_KEY | GEMINI_API_KEY | LLM_API_KEY (openai-compatible, optional)

Model-agnostic by design: prompts ask for JSON in plain words, and callers validate the reply themselves
(`extract_json`). No provider-only feature — tool use, JSON mode, response schemas — is a precondition, so a model
that ignores a formatting hint fails validation instead of passing silently. Keys only ever travel in headers,
never in a URL.
"""
import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request

PROVIDERS = ("anthropic", "openai", "gemini", "openai-compatible")
DEFAULT_MODEL = {"anthropic": "claude-sonnet-5"}
DEFAULT_BASE = {
    "anthropic": "https://api.anthropic.com/v1",
    "openai": "https://api.openai.com/v1",
    "gemini": "https://generativelanguage.googleapis.com/v1beta",
}
KEY_ENV = {"anthropic": "ANTHROPIC_API_KEY", "openai": "OPENAI_API_KEY", "gemini": "GEMINI_API_KEY",
           "openai-compatible": "LLM_API_KEY"}


class ConfigError(ValueError):
    """The environment does not describe a usable endpoint. The message says which variable to set."""


class ProviderError(RuntimeError):
    """The endpoint answered with an error or with a body this adapter cannot read."""


def config_from_env(env=None, default_provider="anthropic"):
    env = os.environ if env is None else env
    provider = (env.get("LLM_PROVIDER") or default_provider).strip().lower()
    if provider not in PROVIDERS:
        raise ConfigError(f"LLM_PROVIDER={provider!r} is not one of {', '.join(PROVIDERS)}")
    model = (env.get("LLM_MODEL") or DEFAULT_MODEL.get(provider) or "").strip()
    if not model:
        raise ConfigError(f"LLM_MODEL is required for {provider} (use a model id from your provider's model list)")
    base = (env.get("LLM_BASE_URL") or DEFAULT_BASE.get(provider) or "").strip().rstrip("/")
    if not base:
        raise ConfigError("LLM_BASE_URL is required for openai-compatible (for Ollama: http://localhost:11434/v1)")
    key = (env.get(KEY_ENV[provider]) or "").strip()
    if not key and provider != "openai-compatible":
        raise ConfigError(f"{KEY_ENV[provider]} is not set")
    return {"provider": provider, "model": model, "base_url": base, "api_key": key}


def build_request(cfg, system, user, max_tokens=2048):
    """Pure: returns (url, headers, body_bytes). No network."""
    p, model, base, key = cfg["provider"], cfg["model"], cfg["base_url"], cfg["api_key"]
    headers = {"content-type": "application/json"}
    if p == "anthropic":
        url = f"{base}/messages"
        headers |= {"x-api-key": key, "anthropic-version": "2023-06-01"}
        body = {"model": model, "max_tokens": max_tokens, "system": system,
                "messages": [{"role": "user", "content": user}]}
    elif p in ("openai", "openai-compatible"):
        url = f"{base}/chat/completions"
        if key:
            headers["authorization"] = f"Bearer {key}"
        body = {"model": model, "messages": [{"role": "system", "content": system}, {"role": "user", "content": user}]}
        # OpenAI's own API deprecated max_tokens (reasoning models reject it); many compatible servers only know
        # max_tokens. Temperature is left at each model's default: several models reject any other value.
        body["max_completion_tokens" if p == "openai" else "max_tokens"] = max_tokens
    elif p == "gemini":
        name = model[len("models/"):] if model.startswith("models/") else model
        url = f"{base}/models/{urllib.parse.quote(name, safe='')}:generateContent"
        headers["x-goog-api-key"] = key
        body = {"systemInstruction": {"parts": [{"text": system}]},
                "contents": [{"role": "user", "parts": [{"text": user}]}],
                "generationConfig": {"maxOutputTokens": max_tokens}}
    else:
        raise ConfigError(f"unknown provider {p!r}")
    return url, headers, json.dumps(body).encode("utf-8")


def parse_response(cfg, data):
    """Returns (text, model) from a provider's JSON reply, or raises ProviderError."""
    p = cfg["provider"]
    try:
        if p == "anthropic":
            text = "".join(c.get("text", "") for c in data["content"] if c.get("type", "text") == "text")
            model = data.get("model") or cfg["model"]
        elif p in ("openai", "openai-compatible"):
            text = data["choices"][0]["message"]["content"] or ""
            model = data.get("model") or cfg["model"]
        else:
            text = "".join(part.get("text", "") for part in data["candidates"][0]["content"]["parts"])
            model = data.get("modelVersion") or cfg["model"]
    except (KeyError, IndexError, TypeError) as e:
        raise ProviderError(f"{p}: unexpected response shape ({type(e).__name__}: {e})") from None
    return text, model


def complete(cfg, system, user, max_tokens=2048, timeout=60, opener=None):
    """One request, no retries. Returns (text, model). Raises ProviderError on HTTP errors and unreadable bodies."""
    url, headers, body = build_request(cfg, system, user, max_tokens)
    req = urllib.request.Request(url, data=body, headers=headers, method="POST")
    open_ = opener or urllib.request.urlopen
    try:
        with open_(req, timeout=timeout) as r:
            raw = r.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        detail = e.read().decode("utf-8", "replace")[:300] if hasattr(e, "read") else ""
        raise ProviderError(f"{cfg['provider']}: HTTP {e.code} {detail}".strip()) from None
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        raise ProviderError(f"{cfg['provider']}: response was not JSON") from None
    return parse_response(cfg, data)


def extract_json(text, kind="array"):
    """First JSON array (or object) in a model reply, tolerating code fences and prose around it."""
    opener, closer = ("[", "]") if kind == "array" else ("{", "}")
    cleaned = re.sub(r"```(?:json)?", "", text)
    start = cleaned.find(opener)
    while start != -1:
        depth, in_str, esc = 0, False, False
        for i in range(start, len(cleaned)):
            ch = cleaned[i]
            if in_str:
                if esc:
                    esc = False
                elif ch == "\\":
                    esc = True
                elif ch == '"':
                    in_str = False
                continue
            if ch == '"':
                in_str = True
            elif ch == opener:
                depth += 1
            elif ch == closer:
                depth -= 1
                if depth == 0:
                    try:
                        return json.loads(cleaned[start:i + 1])
                    except json.JSONDecodeError:
                        break
        start = cleaned.find(opener, start + 1)
    raise ValueError(f"no JSON {kind} in the model reply")


def describe(cfg):
    """Human-readable, key-free description for logs and caches."""
    base = "" if cfg["base_url"] == DEFAULT_BASE.get(cfg["provider"]) else f" @ {cfg['base_url']}"
    return f"{cfg['provider']} · {cfg['model']}{base}"

File: docs/llm.js
/* llm.js — the browser twin of ../llm.py: one small, dependency-free way to call four kinds of LLM endpoint.
   Loaded as a classic <script> (window.LLM) and with require() in node (check-llm.mjs).

   Providers: anthropic (default) · openai · gemini · openai-compatible (Ollama, LM Studio, vLLM, most gateways).
   Model-agnostic: prompts ask for JSON in plain words and callers validate the reply (extractJson). No provider-only
   feature is a precondition. Keys travel in headers only, never in a URL, and are never stored by this file. */
(function (root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory();
  else root.LLM = factory();
})(typeof self !== 'undefined' ? self : this, function () {
  'use strict';
  const PROVIDERS = ['anthropic', 'openai', 'gemini', 'openai-compatible'];
  const LABEL = { anthropic: 'Claude (Anthropic)', openai: 'OpenAI', gemini: 'Google Gemini', 'openai-compatible': 'OpenAI-compatible endpoint' };
  const DEFAULT_MODEL = { anthropic: 'claude-sonnet-5' };
  const DEFAULT_BASE = { anthropic: 'https://api.anthropic.com/v1', openai: 'https://api.openai.com/v1', gemini: 'https://generativelanguage.googleapis.com/v1beta' };

  class ConfigError extends Error {}
  class ProviderError extends Error {}

  function config({ provider = 'anthropic', model = '', baseUrl = '', apiKey = '' } = {}) {
    provider = String(provider).trim().toLowerCase();
    if (!PROVIDERS.includes(provider)) throw new ConfigError(`provider "${provider}" is not one of ${PROVIDERS.join(', ')}`);
    model = String(model || DEFAULT_MODEL[provider] || '').trim();
    if (!model) throw new ConfigError(`a model id is required for ${LABEL[provider]} (use one from your provider's model list)`);
    baseUrl = String(baseUrl || DEFAULT_BASE[provider] || '').trim().replace(/\/+$/, '');
    if (!baseUrl) throw new ConfigError('a base URL is required for an OpenAI-compatible endpoint (for Ollama: http://localhost:11434/v1)');
    apiKey = String(apiKey || '').trim();
    if (!apiKey && provider !== 'openai-compatible') throw new ConfigError(`an API key is required for ${LABEL[provider]}`);
    return { provider, model, baseUrl, apiKey };
  }

  function buildRequest(cfg, system, user, maxTokens = 2048, { browser = false } = {}) {
    const { provider: p, model, baseUrl: base, apiKey: key } = cfg;
    const headers = { 'content-type': 'application/json' };
    let url, body;
    if (p === 'anthropic') {
      url = `${base}/messages`;
      Object.assign(headers, { 'x-api-key': key, 'anthropic-version': '2023-06-01' });
      if (browser) headers['anthropic-dangerous-direct-browser-access'] = 'true';   // Anthropic requires it for calls from a page
      body = { model, max_tokens: maxTokens, system, messages: [{ role: 'user', content: user }] };
    } else if (p === 'openai' || p === 'openai-compatible') {
      url = `${base}/chat/completions`;
      if (key) headers.authorization = `Bearer ${key}`;
      body = { model, messages: [{ role: 'system', content: system }, { role: 'user', content: user }] };
      // OpenAI deprecated max_tokens (reasoning models reject it); many compatible servers only know max_tokens.
      // Temperature stays at each model's default: several models reject any other value.
      body[p === 'openai' ? 'max_completion_tokens' : 'max_tokens'] = maxTokens;
    } else {
      const name = model.startsWith('models/') ? model.slice(7) : model;
      url = `${base}/models/${encodeURIComponent(name)}:generateContent`;
      headers['x-goog-api-key'] = key;
      body = { systemInstruction: { parts: [{ text: system }] }, contents: [{ role: 'user', parts: [{ text: user }] }], generationConfig: { maxOutputTokens: maxTokens } };
    }
    return { url, headers, body: JSON.stringify(body) };
  }

  function parseResponse(cfg, data) {
    try {
      if (cfg.provider === 'anthropic') return { text: data.content.filter(c => (c.type || 'text') === 'text').map(c => c.text || '').join(''), model: data.model || cfg.model };
      if (cfg.provider === 'gemini') return { text: data.candidates[0].content.parts.map(x => x.text || '').join(''), model: data.modelVersion || cfg.model };
      return { text: data.choices[0].message.content || '', model: data.model || cfg.model };
    } catch (e) {
      throw new ProviderError(`${cfg.provider}: unexpected response shape (${e.message})`);
    }
  }

  async function complete(cfg, system, user, { maxTokens = 2048, fetchImpl, browser = typeof window !== 'undefined', timeoutMs = 60000 } = {}) {
    const req = buildRequest(cfg, system, user, maxTokens, { browser });
    const f = fetchImpl || fetch;
    const ctl = typeof AbortController !== 'undefined' ? new AbortController() : null;
    const timer = ctl ? setTimeout(() => ctl.abort(), timeoutMs) : null;
    let res;
    try { res = await f(req.url, { method: 'POST', headers: req.headers, body: req.body, signal: ctl ? ctl.signal : undefined }); }
    catch (e) { throw new ProviderError(`${cfg.provider}: network error (${e.name === 'AbortError' ? 'timed out' : e.message}) — offline, blocked by CORS, or the endpoint is down`); }
    finally { if (timer) clearTimeout(timer); }
    const raw = await res.text();
    if (!res.ok) {
      let detail = raw.slice(0, 300);
      try { const j = JSON.parse(raw); detail = (j.error && (j.error.message || j.error)) || detail; } catch (e) {}
      throw new ProviderError(`${cfg.provider}: HTTP ${res.status} ${typeof detail === 'string' ? detail : JSON.stringify(detail)}`.trim());
    }
    let data;
    try { data = JSON.parse(raw); } catch (e) { throw new ProviderError(`${cfg.provider}: response was not JSON`); }
    return parseResponse(cfg, data);
  }

  function extractJson(text, kind = 'array') {
    const [open, close] = kind === 'array' ? ['[', ']'] : ['{', '}'];
    const s = String(text).replace(/```(?:json)?/g, '');
    let start = s.indexOf(open);
    while (start !== -1) {
      let depth = 0, inStr = false, esc = false;
      for (let i = start; i < s.length; i++) {
        const ch = s[i];
        if (inStr) { if (esc) esc = false; else if (ch === '\\') esc = true; else if (ch === '"') inStr = false; continue; }
        if (ch === '"') inStr = true;
        else if (ch === open) depth++;
        else if (ch === close && --depth === 0) {
          try { return JSON.parse(s.slice(start, i + 1)); } catch (e) { break; }
        }
      }
      start = s.indexOf(open, start + 1);
    }
    throw new Error(`no JSON ${kind} in the model reply`);
  }

  // The settings after the provider is changed: that provider's default model, no base URL and no key. The key is
  // dropped on purpose, so a key typed for one provider is never sent to the next one's endpoint (an OpenAI key to an
  // OpenAI-compatible address run by someone else, say).
  function switchProvider(provider) {
    return { provider, model: DEFAULT_MODEL[provider] || '', baseUrl: '', apiKey: null };
  }

  function describe(cfg) {
    return `${LABEL[cfg.provider]} · ${cfg.model}${cfg.baseUrl === DEFAULT_BASE[cfg.provider] ? '' : ' @ ' + cfg.baseUrl}`;
  }

  return { PROVIDERS, LABEL, DEFAULT_MODEL, DEFAULT_BASE, ConfigError, ProviderError, config, switchProvider, buildRequest, parseResponse, complete, extractJson, describe };
});
